### PR TITLE
Allow the usage of --kubeconfig-dir and --config flags on kubeadm init

### LINKED ISSUE #1403: Cannot specify kubeconfig-dir in kubeadm config
## Is this a BUG REPORT or FEATURE REQUEST?

Choose one: FEATURE REQUEST

When using the `kubeadm init phase kubeconfig` command, we can specify `--kubeconfig-dir` to choose the output directory for the generated kubeconfigs, as in the following example.

```
root@localhost:/tmp/cluster04/pki/kubeconfigs# kubeadm init phase kubeconfig all --kubeconfig-dir $(pwd)/kubeconfigs --cert-dir /tmp/cluster04/pki
[kubeconfig] Using kubeconfig folder "/tmp/cluster04/pki/kubeconfigs/kubeconfigs"
[kubeconfig] Writing "admin.conf" kubeconfig file
[kubeconfig] Writing "kubelet.conf" kubeconfig file
[kubeconfig] Writing "controller-manager.conf" kubeconfig file
[kubeconfig] Writing "scheduler.conf" kubeconfig file
```

If we already have a `kubeadm` config file, we might expect to be able to add the key `kubeconfigDir` to the `ClusterConfiguration` in order to simply pass `--config` to `kubeadm init phase kubeconfig`, such as in the following example `kubeadm` config file.

```
kind: ClusterConfiguration
apiVersion: kubeadm.k8s.io/v1beta1
...
certificatesDir: /tmp/cluster04/pki
kubeconfigDir: /tmp/cluster04/pki/kubeconfigs
clusterName: cluster04
dns:
  type: CoreDNS
imageRepository: k8s.gcr.io
kubernetesVersion: v1.13.3
scheduler: {}
```

However, this does not work, as `kubeconfigDir` is [not a field on `ClusterConfiguration`](https://github.com/kubernetes/kubernetes/blob/v1.14.0-alpha.2/cmd/kubeadm/app/apis/kubeadm/v1beta1/types.go#L60)

```
root@localhost:/tmp/cluster04/pki# kubeadm init phase kubeconfig all --config kubeadm.conf
W0211 22:17:53.619222   22560 strict.go:54] error unmarshaling configuration schema.GroupVersionKind{Group:"kubeadm.k8s.io", Version:"v1beta1", Kind:"ClusterConfiguration"}: error unmarshaling JSON: while decoding JSON: json: unknown field "kubeconfigDir"
[kubeconfig] Using kubeconfig folder "/etc/kubernetes"
```

This results in the kubeconfigs being written to the default directory `/etc/kubernetes` which may be unwanted behavior.

As with all other `kubeadm` commands, we cannot mix `--config` with command line parameters.

```
root@localhost:/tmp/cluster04/pki# kubeadm init phase kubeconfig all --config kubeadm.conf --kubeconfig-dir $(pwd)/kubeconfigs
can not mix '--config' with arguments [kubeconfig-dir]
```

**Request:**
Support a `kubeconfigDir` field in the kubeadm config API such that `kubeadm init phase kubeconfig` can be used with `--config` as expected.

### PR DESCRIPTION
**Pull request by yagonobre**
**What type of PR is this?**
/kind bug
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes/kubeadm/issues/1403

**Special notes for your reviewer**:
I also made a little refactor to ValidateMixedArguments, is it ok?
And renamed from whitelist to allowed

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm: allow the usage of --kubeconfig-dir and --config flags on kubeadm init
```

/priority important-soon
/assign @neolit123

### CODE REVIEW

**Review by neolit123 — APPROVED** [review]
this looks good, thanks @yagonobre 

/retest
/lgtm
/approve
/hold 
@kubernetes/sig-cluster-lifecycle-pr-reviews 

comments from others?

**Comment by fabriziopandini** [comment]
@yagonobre thanks for this PR
I'm ok with this change, considering that as commented on the issues, the `--kubeconfig-dir` remain an additional facility that apply to the kubeconfig phase only. 

/approve
/lgtm

Other changes suggested on the issue (e.g. to promote `--kubeconfig-dir` as a option to be supported by the kubeadm config) IMO require approval from the broader sig before execution.

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (tedyu) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: How about putting these into a Set and check whether flagName is contained in the Set ?

Comment (neolit123) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: our flags are kind of sparse and cover different functionality i.e. the flags don't make sense to be a part of the same set outside of this validatation, but i guess we can consider it for a future refactor.

Comment (tedyu) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: I have a patch ready, should I send out PR ?

Comment (neolit123) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: given the flags are mostly unrelated, my initial reaction is, no.
but you can still send if you want to get feedback on the change.

Comment (tedyu) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: See #74032

**Comment by venomwaqar** [comment]
Sorry I am new to K8s.
I am trying to do exactly same that is generating the kubeconfig files in customer directory rather than /etc/kubernetes/ but kubeadm init not recognising the `--kubeconfig-dir` command, saying unknown flag.

Can anyone please helm me out how do I change the directory?
@yagonobre if you have already done this can you please help me out?

Kubernetes version: 1.13.10
O.S: Centos 7.6

**Comment by neolit123** [comment]
> Can anyone please help me out how do I change the directory?

hi, it is not possible to customize the /etc/kubernetes directory in kubeadm. we don't have plans to make it customizable for the being, but maybe in the future.

in the meantime you can try using a symbolic link pointing to another location.

**Comment by venomwaqar** [comment]
@neolit123 Thanks buddy for letting me know :-)

**Pull request merged**
